File: src/cmd/tools/terminal/ghostty.go
package terminal

// ghostty.go
import (
	"fmt"
	"os"
	"path/filepath"

	"github.com/mpaxson/kettle/src/cmd/helpers"
	"github.com/spf13/cobra"
)

// GhosttyCmd represents the ghostty command
var GhosttyCmd = &cobra.Command{
	Use:   "ghostty",
	Short: "A brief description of your command",
	Long: `A longer description that spans multiple lines and likely contains examples
and usage of using your command. For example:

Cobra is a CLI library for Go that empowers applications.
This application is a tool to generate the needed files
to quickly create a Cobra application.`,
}

var ghosttyInstallCmd = &cobra.Command{
	Use:   "install",
	Short: "Installs ghostty",
	Long:  `Installs ghostty for the current operating system.`,
	Run: func(cmd *cobra.Command, args []string) {
		if !helpers.IsUbuntu() && !helpers.IsDarwin() {
			helpers.PrintFail("Unsupported OS. Only Ubuntu and macOS are supported.")
			return
		}

		if helpers.CommandExists("ghostty") {
			fmt.Println("ghostty is already installed.")
			return
		}

		if helpers.IsDarwin() {
			if !helpers.CommandExists("brew") {
				helpers.PrintFail("Homebrew is not installed. Cannot install ghostty.")
				return
			}
			err := helpers.RunCmd("brew install ghostty")
			if err != nil {
				helpers.PrintFail("Failed to install ghostty with brew")
			}
		} else if helpers.IsUbuntu() {
			err := helpers.RunCmd("sudo snap install ghostty --classic")
			if err != nil {
				helpers.PrintFail("Failed to install ghostty with snap")
			}
		}
	},
}
var ghosttyCreateToggleScriptCmd = &cobra.Command{
	Use:   "create-toggle-script",
	Short: "Creates the ghostty toggle script",
	Long:  `Creates the ghostty toggle script at ~/bin/ghostty-toggle.sh.`,
	Run: func(cmd *cobra.Command, args []string) {
		scriptPath := getScriptPath()
		const scriptContent = `#!/usr/bin/env bash
# Toggle Ghostty as a dropdown terminal (fullscreen on F1)
APP="ghostty"
WM_CLASS="Ghostty"
BASHCMD="bash -c tmux attach || tmux new"
CMD="ghostty --class=$WM_CLASS --fullscreen -e $BASHCMD"
# tdrop options:
# -ma   : match by WM_CLASS
# -w    : window width in %
# -h    : window height in %
# -y    : drop from top (y=0)
tdrop -ma -w 5000 -h 5000 -y 0 -a $CMD`

		// Create bin directory if it doesn't exist
		binDir := filepath.Dir(scriptPath)
		if err := os.MkdirAll(binDir, 0755); err != nil {
			helpers.PrintFail(fmt.Sprintf("Failed to create bin directory: %v", err))
			return
		}

		err := os.WriteFile(scriptPath, []byte(scriptContent), 0755)
		if err != nil {
			helpers.PrintFail(fmt.Sprintf("Failed to create toggle script: %v", err))
			return
		}

		helpers.PrintSuccess(fmt.Sprintf("✅ Created and chmod +x %s", scriptPath))
	},
}

func getScriptPath() string {

	homeDir, err := os.UserHomeDir()
	if err != nil {
		helpers.PrintFail("Failed to get user home directory")
		return ""
	}

	scriptPath := filepath.Join(homeDir, "bin", "ghostty-toggle.sh")
	return scriptPath
}

var ghosttyBindF1Cmd = &cobra.Command{
	Use:   "bind-f1",
	Short: "Bind F1 to toggle Ghostty",
	Long:  `Bind F1 to toggle Ghostty.`,
	Run: func(cmd *cobra.Command, args []string) {
		schema := "org.gnome.settings-daemon.plugins.media-keys"
		path := "/org/gnome/settings-daemon/plugins/media-keys/custom-keybindings/custom0/"
		name := "Ghostty Toggle"
		key := "F1"
		scriptPath := getScriptPath()

		// Check if script exists
		if _, err := os.Stat(scriptPath); os.IsNotExist(err) {
			helpers.PrintFail(fmt.Sprintf("❌ Error: %s not found. Make sure toggle-ghostty exists.", scriptPath))
			return
		}

		// Check if script is executable
		if info, err := os.Stat(scriptPath); err == nil {
			if info.Mode()&0111 == 0 {
				helpers.PrintFail(fmt.Sprintf("❌ Error: %s is not executable. Run: chmod +x %s", scriptPath, scriptPath))
				return
			}
		}

		// Register the custom binding path - this sets the array of custom keybindings
		err := helpers.RunCmd(fmt.Sprintf("gsettings set %s custom-keybindings \"['%s']\"", schema, path))
		if err != nil {
			helpers.PrintFail("Failed to set custom-keybindings")
			return
		}

		// Configure the binding name
		err = helpers.RunCmd(fmt.Sprintf("gsettings set %s.custom-keybinding:%s name \"%s\"", schema, path, name))
		if err != nil {
			helpers.PrintFail("Failed to set binding name")
			return
		}

		// Configure the binding command
		err = helpers.RunCmd(fmt.Sprintf("gsettings set %s.custom-keybinding:%s command \"%s\"", schema, path, scriptPath))
		if err != nil {
			helpers.PrintFail("Failed to set binding command")
			return
		}

		// Configure the key binding
		err = helpers.RunCmd(fmt.Sprintf("gsettings set %s.custom-keybinding:%s binding \"%s\"", schema, path, key))
		if err != nil {
			helpers.PrintFail("Failed to set key binding")
			return
		}

		helpers.PrintSuccess(fmt.Sprintf("✅ Bound %s to %s", key, scriptPath))
	},
}

var ghosttyUnbindF1Cmd = &cobra.Command{
	Use:   "unbind-f1",
	Short: "Remove F1 binding for Ghostty",
	Long:  `Remove F1 binding for Ghostty.`,
	Run: func(cmd *cobra.Command, args []string) {
		const schema = "org.gnome.settings-daemon.plugins.media-keys"
		const path = "/org/gnome/settings-daemon/plugins/media-keys/custom-keybindings/custom0/"
		const key = "F1"
		scriptPath := getScriptPath()

		err := helpers.RunCmd(fmt.Sprintf("gsettings reset %s custom-keybindings", schema))
		if err != nil {
			helpers.PrintFail("Failed to reset custom-keybindings")
			return
		}

		err = helpers.RunCmd(fmt.Sprintf("gsettings reset-recursively %s.custom-keybinding:%s", schema, path))
		if err != nil {
			helpers.PrintFail("Failed to reset-recursively")
			return
		}

		helpers.PrintSuccess(fmt.Sprintf("🗑️  Unbound %s from %s", key, scriptPath))
	},
}

func init() {
	GhosttyCmd.AddCommand(ghosttyInstallCmd)
	GhosttyCmd.AddCommand(ghosttyBindF1Cmd)
	GhosttyCmd.AddCommand(ghosttyUnbindF1Cmd)
	GhosttyCmd.AddCommand(ghosttyCreateToggleScriptCmd)

}

File: src/scripts/tools/terminal/ghostty.py
from invoke import run, task
from scripts.misc import is_ubuntu, is_darwin, console, print, print_success, print_error, config, cmd_exists, is_linux, download_file
import requests
from invoke.context import Context
from pathlib import Path
GITHUB_API = "https://api.github.com/repos/ghostty-org/ghostty/releases/latest"
SCHEMA = "org.gnome.settings-daemon.plugins.media-keys"
PATH = "/org/gnome/settings-daemon/plugins/media-keys/custom-keybindings/custom0/"
SCRIPT = Path.home() / "bin" / "ghostty-toggle.sh"
NAME = "Ghostty Toggle"
KEY = "F1"

from invoke import task
from pathlib import Path
from scripts.misc.helpers import run_sudo
from scripts.tools.general.mac.brew import install as install_brew
def get_asset_url():
    resp = requests.get(GITHUB_API, timeout=10)
    resp.raise_for_status()
    release = resp.json()
    system = config.os.lower()
    urls = []

    for asset in release["assets"]:
        name = asset["name"].lower()
        if is_linux() and name.endswith(".deb"):
            urls.append((asset["browser_download_url"], name))
        elif is_darwin() and (name.endswith(".dmg") or name.endswith(".tar.gz")):
            urls.append((asset["browser_download_url"], name))


    if not urls:
        print_error(f"No suitable Ghostty asset found for {system}")

    return urls[0]  # just grab the first match





@task
def bind_f1(c):
    """
    Bind F1 to toggle Ghostty.
    """
    if not SCRIPT.exists():
        print(f"❌ Error: {SCRIPT} not found. Make sure toggle-ghostty exists.")
        return

    if not SCRIPT.is_file() or not SCRIPT.stat().st_mode & 0o111:
        print(f"❌ Error: {SCRIPT} is not executable. Run: chmod +x {SCRIPT}")
        return

    # Register the custom binding path
    c.run(f"gsettings set {SCHEMA} custom-keybindings \"['{PATH}']\"")

    # Configure the binding
    c.run(f"gsettings set {SCHEMA}.custom-keybinding:{PATH} name \"{NAME}\"")
    c.run(f"gsettings set {SCHEMA}.custom-keybinding:{PATH} command \"{SCRIPT}\"")
    c.run(f"gsettings set {SCHEMA}.custom-keybinding:{PATH} binding \"{KEY}\"")

    print(f"✅ Bound {KEY} to {SCRIPT}")


@task
def unbind_f1(c):
    """
    Remove F1 binding for Ghostty.
    """
    c.run(f"gsettings reset {SCHEMA} custom-keybindings")
    c.run(f"gsettings reset-recursively {SCHEMA}.custom-keybinding:{PATH}")

    print(f"🗑️  Unbound {KEY} from {SCRIPT}")


@task
def install(c):
    c: Context
    if not is_linux() and not is_darwin():
        print_error("Unsupported OS. Only Ubuntu and macOS are supported.")
        return
    
    if cmd_exists('ghostty'):
        print_success("ghostty is already installed.")
        return
    if is_darwin():
        install_brew(c)
        if cmd_exists('brew'):
            c.run("brew install ghostty")
            return
        else:
            print_error("Homebrew is not installed. Cannot install ghostty.")   
            return
            
    elif is_ubuntu():
        run_sudo(f"snap install ghostty --classic")
    else:
        print_error("Unsupported OS. Only Ubuntu and macOS are supported.")

